### PR TITLE
Unanalyzed Fields

### PR DESCRIPTION
**Pull request by adamalix**
This pull request introduces Unanalyzed fields that have the ability to be used as filter queries that are optionally cached in ElasticSearch.

### CODE REVIEW

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 1]
```diff
@@ -782,28 +788,54 @@ trait SolrSchema[M <: Record[M]] extends SlashemSchema[M] {
 
 }
 
+/**
+ * A field type for unanalyzed queries. Results in using Term[V] queries.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala:792]: Maybe add available for ES only.

File: src/test/scala/com/foursquare/slashem/ElasticQueryTest.scala [hunk 2]
```diff
@@ -328,6 +346,21 @@ class ElasticQueryTest extends SpecsMatchers with ScalaCheckMatchers {
     Assert.assertEquals(ids2.intersect(idsWithFavNums).length, 3)
   }
 
+  @Test
+  def testTermQueries {
```
Comment (holdenk) [inline src/test/scala/com/foursquare/slashem/ElasticQueryTest.scala:361]: Add a test for in with empty list, also add a test for nin

File: src/main/scala/com/foursquare/slashem/Schema.scala (hunk 1 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Schema.scala:792]: Then do we want separate SlashemGeoFields for Solr / ES?  This should behave the same for ES / Solr with the current impl.

Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala:792]: If SlashemGeoFields generate Term[V] queries, then Term[V] queries need to
have query-gen support for Solr.

On Thu, May 10, 2012 at 5:32 PM, Adam Alix <
reply@reply.github.com

> wrote:
> 
> > @@ -782,28 +788,54 @@ trait SolrSchema[M <: Record[M]] extends
> > SlashemSchema[M] {
> > 
> >  }
> > 
> > +/**
> > - \* A field type for unanalyzed queries. Results in using Term[V]
> >   queries.
> 
> Then do we want separate SlashemGeoFields for Solr / ES?  This should
> behave the same for ES / Solr with the current impl.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/foursquare/slashem/pull/34/files#r806027

## 

Cell : 425-233-8271

Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Schema.scala:792]: It does, you're not seeing the latest because of the way this thing does reviews.

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 3]
```diff
@@ -405,6 +408,47 @@ object Ast {
...
+  case class Term[T](query: Iterable[T], escaped: Boolean = true, cached: Boolean = true) extends Query[T] {
+    // hack for single term queries
+    def this(query: T) = this(List(query))
+    /** @inheritdoc */
+    //def extend() = throw new UnimplementedException("Slashem does not support Term queries Solr")
+    def extend(): String = {
+      escaped match {
+        // hack to fix wrapping the queries in a List()
+        case true => {'"' + escape(query.mkString("")) + '"'}
+        case false => '"' + query.toString + '"'
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Don't you want to do query.mkString("") here to?
Also add a test case for single element Term[T] query against Solr.

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 4]
```diff
@@ -405,6 +408,47 @@ object Ast {
...
+    }
+    /** @inheritdoc */
+    def elasticExtend(qf: List[WeightedField], pf: List[PhraseWeightedField], mm: Option[String]): ElasticQueryBuilder = {
+      val fieldName = qf.head.fieldName
+      val weight = qf.head.weight.toFloat
+      query match {
+        case term::Nil => EQueryBuilders.termQuery(fieldName, term).boost(weight)
+        case terms => {
+          val moarTerms = terms.toSeq.map(_.toString)
+          EQueryBuilders.termsQuery(fieldName, moarTerms: _*).boost(weight)
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala:440]: Is this going to work with an empty terms? Or do we need to special case it?

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 5]
```diff
@@ -405,6 +408,47 @@ object Ast {
...
+      }
+    }
+    /** @inheritdoc */
+    override def elasticFilter(qf: List[WeightedField]): ElasticFilterBuilder = {
+      val fieldName = qf.head.fieldName
+      query match {
+        case term::Nil => EFilterBuilders.termFilter(fieldName, term).cache(cached)
+        case terms => {
+          val moarTerms = terms.toSeq.map(_.toString)
+          EFilterBuilders.termsFilter(fieldName, moarTerms: _*).cache(cached)
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala:451]: Is this going to work with an empty terms? Or do we need to special case it?

File: src/main/scala/com/foursquare/slashem/Schema.scala (hunk 1 above)
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala:792]: Fuck github. Sorry for the dumb comment :p

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 6]
```diff
@@ -843,6 +875,14 @@ trait SlashemField[V, M <: Record[M]] extends OwnedField[M] {
 
 //Slashem field types
 class SlashemStringField[T <: Record[T]](owner: T) extends StringField[T](owner, 0) with SlashemField[String, T]
+/**
+ * Field type that can be queried without analyzing whitespace.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: I'd recommend "Field type that be queried without analyzing. An example of this would be a multi-value field or a whitespace tokenized field where search terms are always for a specific token". What do you think?

**Comment by adamalix** [comment]
Deleted branch because it got corrupt. Opening new request.
